File: basics/account-data/native/program/src/state/address_info.rs
use borsh::{BorshDeserialize, BorshSerialize};

#[derive(BorshDeserialize, BorshSerialize, Debug)]
pub struct AddressInfo {
    pub name: String,
    pub house_number: u8,
    pub street: String,
    pub city: String,
}

impl AddressInfo {
    pub fn new(name: String, house_number: u8, street: String, city: String) -> Self {
        AddressInfo {
            name,
            house_number,
            street,
            city,
        }
    }
}

File: basics/realloc/native/ts/state/address-info.ts
import { Buffer } from 'node:buffer';
import * as borsh from 'borsh';

export class AddressInfo {
  name: string;
  house_number: number;
  street: string;
  city: string;

  constructor(props: {
    name: string;
    house_number: number;
    street: string;
    city: string;
  }) {
    this.name = props.name;
    this.house_number = props.house_number;
    this.street = props.street;
    this.city = props.city;
  }

  toBase58() {
    return borsh.serialize(AddressInfoSchema, this).toString();
  }

  toBuffer() {
    return Buffer.from(borsh.serialize(AddressInfoSchema, this));
  }

  static fromBuffer(buffer: Buffer) {
    return borsh.deserialize(AddressInfoSchema, AddressInfo, buffer);
  }
}

export const AddressInfoSchema = new Map([
  [
    AddressInfo,
    {
      kind: 'struct',
      fields: [
        ['name', 'string'],
        ['house_number', 'u8'],
        ['street', 'string'],
        ['city', 'string'],
      ],
    },
  ],
]);
